### ISSUE TITLE
Add support for setting per-service variables (multi-container / multi-service)

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#1153]
Probably using a syntax like:
`balena env add varName varValue --application myApp --service myService`

but the exact syntax / command may be decided later.

**Comment by pdcastro** [comment]
Vaguely related issues:
* [Cannot SSH to container to MC application](https://github.com/balena-io/balena-cli/issues/856)
* [Support multi-container local push](https://github.com/balena-io/balena-cli/issues/836)
* [Update docs to reflect MC support](https://github.com/balena-io/balena-cli/issues/806)
* [document that balena sync doesn't work with multicontainer](https://github.com/balena-io/balena-cli/issues/1139)

**Comment by magiconair** [comment]
Please focus on getting feature parity with the UI first. You don't have to solve all the multi-container issues at once. It is possible to set a service env via the UI so just expose that feature via the CLI through an extra cmd line arg.

**Comment by magiconair** [comment]
We're working around this by prefixing config variables for different services, e.g. `FOO_CONFIG`. However, this has the effect that when we update the configuration for a single service, all services are restarted.

**Comment by magiconair** [comment]
@pdcastro Any update here?

**Comment by pdcastro** [comment]
@magiconair, thanks for checking on this issue's progress. This issue has indeed been discussed in a backlog priority meeting just a few days ago, and work on it should start shortly (hopefully next week, if not then soon after).

**Comment by pdcastro** [comment]
Status update: some work has been done on "strategic pre-requisites" for this issue (PR #1413). I then came across a balena-ci (our continuous integration system) issue that is blocking further progress, but I expect to resolve that CI issue one way or another this week.

**Comment by magiconair** [comment]
@pdcastro Any update? You must have an API that the browser uses to perform this function. Can't you just hook this up to the CLI until your API refactor is ready? It's been more than 6 months to get some feature parity between the UI and the CLI. How are we supposed to script and automate with Balena if we cannot use integrate the functions the platform offers?

**Comment by pdcastro** [comment]
@magiconair, sorry for the delay. The refactor was done and I don't think there is anything else in the way of implementing this issue -- it has been a matter of juggling requests and priorities from multiple sides. Two pieces of good news though: an additional engineer joined the effort on the CLI project earlier this month (hi @srlowe!), :-) and I am currently reviewing all open issues in the CLI repo and sorting them by priority in this [Sorted Backlog](https://github.com/balena-io/balena-cli/milestone/2) (by the way, it is meant to be public, let me know in case you don't have access). I am still adding issues to it, but as it stands, this issue is very close to the top, so we should get to it sooner than later. Thank you for your feedback and patience.

**Comment by pdcastro** [comment]
This feature was (finally...) implemented in CLI v11.20.0: [releases](https://github.com/balena-io/balena-cli/releases). For usage information, check the help output of the commands:
```
balena help envs
balena help env add
balena help env rm
balena help env rename
```
The changes were conservative to maintain backwards compatibility in printed output, but a few new flags were introduced to provide additional information and facilitate scripting, in particular:
* The `--all` option to `balena envs` prints additional columns in the tabular output, to discriminate between variable scope: application (fleet), device or service.
* The `-j` or `--json` option to `balena envs` causes the output to be printed in JSON format, and produces an empty JSON array instead of an error message when no variables exist for the given query.
* And, of course, the `--service` option applies to all `env` commands to select / indicate variables in service scope. This flag takes a service name as argument in the case of the `envs` and `env add` command, but it is a boolean flag (no arguments) in the case of the `rm` and `rename` commands, which identify the variable by its numeric database ID.

Any feedback would be most welcome! I am closing this issue to indicate it is completed, but it is still possible to add comments. Thank you for the feature request and patience while this issue made its way through the backlog.
